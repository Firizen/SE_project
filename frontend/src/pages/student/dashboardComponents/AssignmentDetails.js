import { useState, useEffect } from "react";
import { FaArrowLeft } from "react-icons/fa";

function AssignmentDetails({ assignment, resetSelection }) {
  const [submitted, setSubmitted] = useState(null);
  const [selectedFile, setSelectedFile] = useState(null);
  const [filePreview, setFilePreview] = useState(null);
  const [uploadedFileURL, setUploadedFileURL] = useState(null);
  const [uploadedFileType, setUploadedFileType] = useState(null);

  useEffect(() => {
    const studentID = localStorage.getItem("studentID");
    if (!studentID) {
      alert("Student not logged in!");
      return;
    }

    fetch(`http://localhost:5000/api/submissions/checkStudentSubmission?studentID=${studentID}&assignmentID=${assignment._id}`)
      .then((res) => res.json())
      .then((data) => {
        setSubmitted(data.submitted);
        if (data.submitted) {
          fetchSubmittedDocument(studentID, assignment._id);
        }
      })
      .catch((err) => console.error("Error checking submission:", err));
  }, [assignment._id]);

  const fetchSubmittedDocument = async (studentID, assignmentID) => {
    try {
      const res = await fetch(`http://localhost:5000/api/submissions/getSubmittedDocument?studentID=${studentID}&assignmentID=${assignmentID}`);
      if (!res.ok) throw new Error("No uploaded document found");

      const contentType = res.headers.get("Content-Type");
      setUploadedFileType(contentType);
      const blob = await res.blob();
      setUploadedFileURL(URL.createObjectURL(blob));
    } catch (error) {
      console.error("Error fetching submitted document:", error);
      alert("Failed to fetch document");
    }
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    setSelectedFile(file);

    if (file) {
      const fileURL = URL.createObjectURL(file);
      setFilePreview(fileURL);
    }
  };

  const handleSubmit = async () => {
    if (!selectedFile) {
      alert("Please select a file before submitting.");
      return;
    }

    const studentID = localStorage.getItem("studentID");
    if (!studentID) {
      alert("Student not logged in!");
      return;
    }

    const formData = new FormData();
    formData.append("studentID", studentID);
    formData.append("assignmentID", assignment._id);
    formData.append("document", selectedFile);

    try {
      const response = await fetch("http://localhost:5000/api/submissions/submit", {
        method: "POST",
        body: formData
      });

      if (!response.ok) {
        throw new Error("Failed to submit assignment");
      }

      alert("Assignment submitted successfully!");
      setSubmitted(true);
      fetchSubmittedDocument(studentID, assignment._id);
      setFilePreview(null); // Clear preview after submission
    } catch (err) {
      console.error("Error submitting assignment:", err);
      alert("Submission failed!");
    }
  };

  return (
    <div className="p-8 w-full h-full bg-white flex">
      <div className="w-1/3 p-4">
        <button onClick={resetSelection} className="text-gray-600 text-2xl mb-4">
          <FaArrowLeft />
        </button>

        <h3 className="text-2xl font-semibold mb-4">{assignment.title}</h3>
        <p className="text-gray-600">
          <strong>Created by:</strong> {assignment.teacherName || "Unknown"}
        </p>
        <p className="text-gray-600">
          <strong>Description:</strong> {assignment.description}
        </p>
        <p className="text-gray-600">
          <strong>Due Date:</strong> {new Date(assignment.dueDate).toLocaleString()}
        </p>

        {!submitted && (
          <div className="mt-4">
            <label className="block text-gray-700">Upload Document:</label>
            <input type="file" onChange={handleFileChange} className="border p-2 w-full" />
          </div>
        )}

        {submitted ? (
          <button className="bg-gray-400 mt-6 text-white px-4 py-2 rounded-md shadow cursor-not-allowed" disabled>
            Assignment Submitted
          </button>
        ) : (
          <button
            onClick={handleSubmit}
            className="bg-blue-500 mt-6 text-white px-4 py-2 rounded-md shadow hover:bg-blue-600"
          >
            Submit
          </button>
        )}
      </div>

      <div className="w-2/3 border-l p-4 h-full flex justify-center items-center">
        {filePreview ? (
          <div className="w-full h-full">
            <h4 className="text-lg font-semibold mb-2">Selected Document Preview:</h4>
            {selectedFile.type === "application/pdf" ? (
              <iframe src={filePreview} className="w-full h-full border" title="PDF Preview"></iframe>
            ) : selectedFile.type.startsWith("image/") ? (
              <img src={filePreview} alt="Selected File" className="w-full h-auto" />
            ) : (
              <p className="text-gray-600">Preview not available for this file type.</p>
            )}
          </div>
        ) : uploadedFileURL ? (
          <div className="w-full h-full">
            <h4 className="text-lg font-semibold mb-2">Uploaded Document:</h4>
            {uploadedFileType === "application/pdf" ? (
              <iframe src={uploadedFileURL} className="w-full h-full border" title="PDF Preview"></iframe>
            ) : uploadedFileType?.startsWith("image/") ? (
              <img src={uploadedFileURL} alt="Submitted File" className="w-full h-auto" />
            ) : (
              <p className="text-gray-600">Preview not available for this file type.</p>
            )}
          </div>
        ) : (
          <p className="text-gray-600">No uploaded document found.</p>
        )}
      </div>
    </div>
  );
}

export default AssignmentDetails;
