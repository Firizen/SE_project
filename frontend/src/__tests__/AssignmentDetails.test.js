import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AssignmentDetails from "../pages/student/dashboardComponents/AssignmentDetails";

test("renders AssignmentDetails correctly", () => {
  const mockAssignment = {
    _id: "123",
    title: "Sample Assignment",
  };

  render(<AssignmentDetails assignment={mockAssignment} />);
  expect(screen.getByText("Sample Assignment")).toBeInTheDocument();
});

test("displays 'No assignment selected' if no assignment is provided", () => {
  render(<AssignmentDetails assignment={null} />);
  expect(screen.getByText("No assignment selected")).toBeInTheDocument();
});

test("calls resetSelection when the back button is clicked", async () => {
  const mockAssignment = { _id: "123", title: "Sample Assignment" };
  const resetSelection = jest.fn();

  render(<AssignmentDetails assignment={mockAssignment} resetSelection={resetSelection} />);
  
  const backButton = screen.getByRole("button", { name: /back/i });
  await userEvent.click(backButton);

  expect(resetSelection).toHaveBeenCalledTimes(1);
});

test("matches snapshot", () => {
  const mockAssignment = { _id: "123", title: "Sample Assignment" };
  const { asFragment } = render(<AssignmentDetails assignment={mockAssignment} />);
  expect(asFragment()).toMatchSnapshot();
});
