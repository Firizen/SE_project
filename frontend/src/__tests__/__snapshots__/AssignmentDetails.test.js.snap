// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-8 w-full h-full bg-white flex"
  >
    <div
      class="w-1/3 p-4"
    >
      <button
        class="text-gray-600 text-2xl mb-4"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      <h3
        class="text-2xl font-semibold mb-4"
      >
        Sample Assignment
      </h3>
      <p
        class="text-gray-600"
      >
        <strong>
          Created by:
        </strong>
         Unknown
      </p>
      <p
        class="text-gray-600"
      >
        <strong>
          Description:
        </strong>
         
      </p>
      <p
        class="text-gray-600"
      >
        <strong>
          Due Date:
        </strong>
         Invalid Date
      </p>
      <div
        class="mt-4"
      >
        <label
          class="block text-gray-700"
        >
          Upload Document:
        </label>
        <input
          class="border p-2 w-full"
          type="file"
        />
      </div>
      <button
        class="bg-blue-500 mt-6 text-white px-4 py-2 rounded-md shadow hover:bg-blue-600"
      >
        Submit
      </button>
    </div>
    <div
      class="w-2/3 border-l p-4 h-full flex justify-center items-center"
    >
      <p
        class="text-gray-600"
      >
        No uploaded document found.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
